org.apache.catalina.startup.ContextConfig

delcroip opened the pull request:
To be able to set the dhis2 home from a context var instead of system property so one can have multiple DHIS2 running without messing the env variable (and running multiple tomcat)

Solution inspired from https://stackoverflow.com/questions/372686/how-can-i-specify-system-properties-in-tomcat-configuration-on-startup

Example of context file to put on TOMCAT_HOME/conf/Catalina/localhost/<warname>.xml when using Catalina engine and no specific host configured


```
<Context>

  <Environment name="dhis2-home" value="/home/dhis/config-2"
         type="java.lang.String" override="false"/>

</Context>
```

I tested the solution, DHIS2 detect the right value

stian-sandvold commented on dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/location/DefaultLocationManager.java:
```diff
@@ -96,12 +106,18 @@ public void init()
+                Context initCtx = new InitialContext();
+                Context envCtx = ( Context ) initCtx.lookup( "java:comp/env" );
+                path = ( String ) envCtx.lookup( this.contextVariable );
+            }catch ( NamingException e ) 
```
```suggestion
            }
            catch ( NamingException e ) 
```

delcroip commented:
Dear @stian-sandvold ,

Is there any issue with that PR, If yes can I help ?

Thanks in advance

br

larshelge commented:
Thanks @delcroip for the patch. Just checking, is there any change in existing behavior here, i.e. would the system still read the DHIS 2  environment variable and system property as before? Which order/priority has the context variable?

delcroip commented:
Hello @larshelge 
the order is this one:
1- systemProperty 
2- context variable
3- Env variable

The reason is that in many case the Env variable is used to configure DHIS2; The goal of this change it to have several dhis2 on the same tomcat, so the context variable, when set, should have a priority in order to not mess with the existing DHIS2 install (if any)

best regards

delcroip commented:
Hello @larshelge ,

I tried to solve a merge issue but it seems I don't respect some rules, the message from sonarcloud is not clear, can you tell me what I did wrong ? (I didn't change anything but add your import which was creating the conflict)
 br

stian-sandvold commented:
Hey delcroip,

sorry for the slow response.

I had another look at your PR, and have some feedback:

1. The priority order of the different variables looks good. Looks like the more specific options take priority.
2. You are swallowing an exception in your PR. It would be great to either log (debug) and/or put a comment explaining why we are swallowing it, and what it implies.
3. The failing codestyle check is most likely due to the 4 newlines you added on lines 93-96. I don't see any other style-issues just by looking at it.
4. I do not know exactly what the 3 critical issues from SonarQube is, since the report is not available. But I suspect it's related to swallowing the exception and code complexity of the method. I suspect my problems re-running it is due to the PR residing in your repository and not the dhis2 one.

To me, I'm happy with merging this as long as you fix item 2. If you have ideas to reduce complexity they are also welcome.

delcroip commented:
@stian-sandvold 
I changed the priority because the context variable is more specific than the system variable
I addressed point 2,
I remove the useless lines
I tried to reduce code complexity and make the code more readable

I tested (during maven build) with default, system var and Env var, I didn't retest the contex variable because I couldn't with maven, but I will do it a bit later
